### PR TITLE
Update unbacked symints in masked_select more precisely

### PR DESCRIPTION
**Pull request by jackzhxng**
## Summary
At the moment, the fake impl for `masked_select` simply sets the upper range while updating its size-like SymInt to `sys.maxsize`(9223372036854775807, max value for an unsigned int64) if the there are any SymInts in the original input tensor shape. This PR constrains the range more intelligently by using the upper ranges of each SymInt in the input tensor shape.

This solves an issue where a model being lowered to Executorch errors during memory planning because the memory allocated for `masked_select` ended up exceeded the 64-bit address space (`INT_MAX * size(dtype)`).

## Test plan
- Passes existing unit tests (tests case where upper bound is inf)
- Added unit test to verify upper bound reduction calculation
- Tested end-to-end by exporting with TORCH_LOGS="export" and ensuring that the range for `masked_select`'s SymInt size has the correct upper bound

### CODE REVIEW

**Review by ezyang — CHANGES_REQUESTED** [review]
Needs test

File: torch/_subclasses/fake_impls.py [hunk 1]
```diff
@@ -455,9 +457,28 @@ def masked_select(fake_mode, func, self, mask):
...
+            for size in self.shape:
+                if isinstance(size, int):
+                    num_elements *= size
+                elif isinstance(size, torch.SymInt):
+                    sym_node = size.node
+                    sym_range = bound_sympy(
+                        sym_node.expr, sym_node.shape_env.var_to_range
+                    )
+                    if isinstance(sym_range.upper, IntInfinity):
+                        num_elements = sys.maxsize - 1
```
Comment (ezyang) [inline torch/_subclasses/fake_impls.py:477]: Why don't you just use int infinity here?

File: torch/_subclasses/fake_impls.py [hunk 2]
```diff
@@ -455,9 +457,28 @@ def masked_select(fake_mode, func, self, mask):
...
+                        break
+                    num_elements *= int(sym_range.upper)
```
Comment (ezyang) [inline torch/_subclasses/fake_impls.py:479]: I think you do not have to do it this manually. Instead, do math.prod(self.shape) and then read out the var_to_range upper bound from the result. You can also generate a minimum bound as well this way.

File: torch/_subclasses/fake_impls.py (hunk 1 above)
Comment (dvorjackz) [reply torch/_subclasses/fake_impls.py:477]: Keeping consistent with https://github.com/pytorch/pytorch/blob/ea89f012817d54a1b858d8f62f0f3deea27d008a/torch/_subclasses/fake_impls.py#L450

File: torch/_subclasses/fake_impls.py (hunk 2 above)
Comment (dvorjackz) [reply torch/_subclasses/fake_impls.py:479]: Hmm what should be the code here? I'm getting the following output on the example in my unit test, not sure what to do with these.

```
>> math.prod(self.shape)
5*s0*s1
>> math.prod(self.shape).node.shape_env.var_to_range
{s0: VR[2, 100], s1: VR[2, 7], u0: VR[-int_oo, int_oo]}
```

Comment (ezyang) [reply torch/_subclasses/fake_impls.py:479]: You still have to use bound_sympy

File: torch/_subclasses/fake_impls.py (hunk 1 above)
Comment (ezyang) [reply torch/_subclasses/fake_impls.py]: Is out of date but no matter, can be updated separately

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**